README.md
# CTAN submit

[![CI tests status badge][ci-shield]][ci-url]

[![Latest release badge][release-shield]][release-url]
[![License badge][license-shield]][license-url]

[ci-url]: https://github.com/paolobrasolin/ctan-submit-action/actions/workflows/main.yml "CI tests"
[ci-shield]: https://img.shields.io/github/actions/workflow/status/paolobrasolin/ctan-submit-action/main.yml?branch=main&label=nightly%20check&logo=github
[release-url]: https://github.com/paolobrasolin/ctan-submit-action/releases "Latest release"
[release-shield]: https://img.shields.io/github/v/release/paolobrasolin/ctan-submit-action?display_name=tag&sort=semver
[license-url]: https://github.com/paolobrasolin/ctan-submit-action/blob/main/LICENSE "License"
[license-shield]: https://img.shields.io/github/license/paolobrasolin/ctan-submit-action

This action performs a submit on the CTAN API to validate or upload a package.

## Usage

As any other action, it's recommended to use the latest stable version: `paolobrasolin/ctan-submit-action@v1`.

### Inputs

#### `version`

This is the CTAN API version to address.

This input is **not required** and has **no default**.

Effectively, the latest CTAN API version will be addressed by default.
That is most probably what you want.

Currently versions `1.1` and `1.0` (deprecated) exist.
More details at [Submitting to CTAN](https://ctan.org/help/submit/).

#### `action`

This is the action to perform upon submit.

This input is **not required** and **defaults to `validate`**.

Allowed values are:

- `validate` to validate your submission.
  More details at [Sending a Validation Request to CTAN
  ](https://ctan.org/help/submit/#validation)

- `upload` to validate and upload your submission.
  More details at [Sending a Submission Request to CTAN
  ](https://ctan.org/help/submit/#submit)

#### `file_path`

This is the archive file path of the package you are submitting.

This input is **required** and has **no default**.

You can find more details on structuring the archive at [How can I upload a package?
](https://ctan.org/help/upload-pkg) and [Additional Information for CTAN Uploaders
](https://ctan.org/file/help/ctan/CTAN-upload-addendum).

#### `fields`

This is a YAML string containing all metadata of the package you are submitting.

This input is **required** and has **no default**.

You can find exhaustive documentation for the available parameters at [Parameters for Validation](https://ctan.org/help/submit#validation.parameters).

### Outputs

Currently there are no outputs.

The action will succeed (fail) if the request to the API succeeds (fails).

All errors, warnings and informative messages given by CTAN will be displayed in the log of your job.

## Example workflows

The implementation details of your workflow will depend very much on your needs, but here a few examples you can get inspiration from.

### Validating

There is a very simple workflow running a nightly check on this repo to ensure everything is in working order:

[![CI tests status badge][ci-shield]][ci-url]

It uses CTAN API to perform a simple validity check, without actually submitting anything.
As long as the badge is green we can sleep easy using this action.

You can read its [full sourcode](https://github.com/paolobrasolin/ctan-submit-action/blob/main/.github/workflows/main.yml), but here's the gist of its structure:

```yaml
name: validate
on:
  # ...
jobs:
  post:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout repo
        uses: actions/checkout@v2
      - name: CTAN submit (validate only)
        uses: paolobrasolin/ctan-submit-action@v1
        with:
          action: validate
          # We use the file_path of a pre-built test fixture committed to the repo. You'd normally use the path of the archive resulting from your build steps.
          file_path: ${{ github.workspace }}/test/fixtures/htunk.zip
          fields: |
            pkg: htunk
            version: 0.1.0
            author: Paolo Brasolin
            email: paolo.brasolin@gmail.com 
            uploader: paolo.brasolin@gmail.com 
            license: mit
            summary: This is a CI test
```

### Releasing

Normally you'd want to use this action to publish releases to CTAN (and perhaps to GitHub itself).

Here are the broad strokes of how you could implement that:

```yaml
name: release
on:
  push:
    tags:
      # We match pushes of semantic versioning tags.
      - v[0-9]+.[0-9]+.[0-9]+
jobs:
  release:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout repo
        uses: actions/checkout@v2

      - # This step extracts the tag and the version into variables for reuse; you can do this in your preferred way.
        name: Labeller
        id: labeller
        run: |
          echo "TAG=${GITHUB_REF/refs\/tags\//}" >> $GITHUB_OUTPUT
          echo "VERSION=${GITHUB_REF/refs\/tags\/v/}" >> $GITHUB_OUTPUT

      - # This step is where you build the distributable of your package, named foobar.
        name: Build
        run: |
          # TODO: somehow build and archive your package into dist/foobar.zip
          cp dist/foobar.zip dist/foobar-${{ steps.labeller.outputs.version }}.zip

      - # Optionally, you can also publish your release to GitHub.
        name: Release on Github
        uses: docker://antonyurchenko/git-release:v3
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          DRAFT_RELEASE: "false"
          CHANGELOG_FILE: "CHANGELOG.md"
          ALLOW_EMPTY_CHANGELOG: "false"
          ALLOW_TAG_PREFIX: "true"
          RELEASE_NAME: "Foobar ${{ steps.labeller.outputs.version }}"
        with:
          args: |
            dist/foobar-${{ steps.labeller.outputs.version }}.zip

      - # This step is executed only on the first submission of your package to CTAN, assuming your first release tag is v1.0.0. Note that it requires a lot more fields than the simple update step below. Please read and follow CTAN documentation to fill out everything correctly.
        name: Submit to CTAN (first release)
        if: ${{ steps.labeller.outputs.version == '1.0.0' }}
        uses: paolobrasolin/ctan-submit-action@v1
        with:
          action: upload
          file_path: dist/foobar-${{ steps.labeller.outputs.version }}.zip
          fields: |
            update: "false"
            pkg: foobar
            version: ${{ steps.labeller.outputs.version }}
            uploader: Paolo Brasolin
            email: paolo.brasolin@gmail.com
            author: Paolo Brasolin
            license: mit
            ctanPath: /graphics/pgf/contrib/foobar
            topic: diagram
            home: https://paolobrasolin.github.io/foobar/
            bugs: https://github.com/paolobrasolin/foobar/issues
            repository: https://github.com/paolobrasolin/foobar
            summary: >
              Foobar: foobars for TeX
            description: >
              Foobar is a TikZ library. Its aim is making foobars more approachable for people using TikZ to draw them.
            note: >
              Extra relevant topics are: diagram-maths maths pgf-tikz

      - # This step is executed on all subsequent submissions of your package to CTAN. In fact, this is simply an update to an existing package.
        name: Submit to CTAN (update)
        if: ${{ steps.labeller.outputs.version != '1.0.0' }}
        uses: paolobrasolin/ctan-submit-action@v1
        with:
          action: upload
          file_path: dist/foobar-${{ steps.labeller.outputs.version }}.zip
          fields: |
            update: "true"
            pkg: foobar
            version: ${{ steps.labeller.outputs.version }}
            uploader: Paolo Brasolin
            email: paolo.brasolin@gmail.com
```

### Other examples

Here is the list of all projects using this action: https://github.com/paolobrasolin/ctan-submit-action/network/dependents

You can peek at their CI setup to find other live examples and learn more about how it can be used.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]

### Added

- Commented examples in readme.

### Fixed

- Freshen up broken links in readme.

## [1.0.0] - 2021-11-20

### Added

- First public release.

[unreleased]: https://github.com/paolobrasolin/ctan-submit-action/compare/v1.0.0...HEAD
[1.0.1]: https://github.com/paolobrasolin/ctan-submit-action/compare/v1.0.0...v1.0.1
[1.0.0]: https://github.com/paolobrasolin/ctan-submit-action/releases/tag/v1.0.0


run.test.js
const YAML = require("yaml");

const run = require("./run");

describe("run", () => {
  const ENV_BAK = process.env;

  beforeEach(() => {
    process.exitCode = undefined;
    process.env = { ...ENV_BAK };
  });

  test("failure (missing inputs)", async () => {
    expect(process.exitCode).toBe(undefined);
    await run();
    expect(process.exitCode).toBe(1);
  });

  test("failure (invalid inputs)", async () => {
    process.env["INPUT_VERSION"] = "1.0";
    process.env["INPUT_ACTION"] = "validate";
    process.env["INPUT_FILE_PATH"] = "htunk.zip";
    process.env["INPUT_FIELDS"] = YAML.stringify({
      pkg: "knuth",
      version: "foobar",
      author: "foobar",
      email: "foobar",
      uploader: "foobar",
      license: "bsd",
      summary: "foobar"
    });
    expect(process.exitCode).toBe(undefined);
    await run();
    expect(process.exitCode).toBe(1);
  });

  test("success", async () => {
    process.env["INPUT_VERSION"] = "1.0";
    process.env["INPUT_ACTION"] = "validate";
    process.env["INPUT_FILE_PATH"] = "htunk.zip";
    process.env["INPUT_FIELDS"] = YAML.stringify({
      pkg: "htunk",
      version: "foobar",
      author: "foobar",
      email: "foobar",
      uploader: "foobar",
      license: "bsd",
      summary: "foobar"
    });
    expect(process.exitCode).toBe(undefined);
    await run();
    expect(process.exitCode).toBe(undefined);
  });
});


ctan.test.js
const rewire = require("rewire");

// NOTE: we're just testing the internals
const CTAN = rewire("./ctan");

describe("buildUrl", () => {
  const buildUrl = CTAN.__get__("buildUrl");

  test("default url is endpoint for validation on latest version", () => {
    expect(buildUrl()).toBe("https://www.ctan.org/submit/validate");
  });

  test("version can be customized", () => {
    expect(buildUrl({ version: "2.0" })).toBe(
      "https://www.ctan.org/submit/2.0/validate"
    );
  });

  test("action can be customized", () => {
    expect(buildUrl({ action: "upload" })).toBe(
      "https://www.ctan.org/submit/upload"
    );
  });

  test("everything can be customized", () => {
    expect(
      buildUrl({
        version: "2.0",
        action: "upload"
      })
    ).toBe("https://www.ctan.org/submit/2.0/upload");
  });
});

describe("postCallback", () => {
  const postCallback = CTAN.__get__("postCallback");

  test("throws request error if present", () => {
    expect(() => {
      postCallback(null, "This is an error!", null, null);
    }).toThrow("This is an error!");
  });

  test("logs response code", () => {
    const logger = { info: jest.fn() };
    postCallback(logger, null, 200, null);
    expect(logger.info).toHaveBeenCalledWith("CTAN responded with code 200");
  });

  test("logs body messages", () => {
    const logger = { info: jest.fn(), warning: jest.fn() };
    postCallback(logger, null, 200, '[["WARNING", "Label", "Details"]]');
    expect(logger.info).toHaveBeenCalledWith("CTAN responded with code 200");
    expect(logger.warning).toHaveBeenCalledTimes(1);
  });

  test("throws error if status code is failure", () => {
    const logger = { info: jest.fn() };
    expect(() => {
      postCallback(logger, null, 500, null);
    }).toThrow("CTAN submit failed: see log for details");
  });
});

describe("logBody", () => {
  const logBody = CTAN.__get__("logBody");

  test("passes through non-JSON body", () => {
    const logger = { info: jest.fn() };
    logBody(logger, "foobar");
    expect(logger.info).toHaveBeenCalledWith("CTAN says foobar");
  });

  test("passes through non-list JSON body", () => {
    const logger = { info: jest.fn() };
    logBody(logger, '{"foo": "bar"}');
    expect(logger.info).toHaveBeenCalledWith('CTAN says {"foo": "bar"}');
  });

  test("pretty prints a list JSON body", () => {
    const logger = { info: jest.fn() };
    logBody(logger, '[["INFO", "Label", "Details"]]');
    expect(logger.info).toHaveBeenCalledWith(
      'CTAN says ["INFO","Label","Details"]'
    );
  });
});

describe("maybeJSONParse", () => {
  const maybeJSONParse = CTAN.__get__("maybeJSONParse");

  test("parses valid JSON", () => {
    expect(maybeJSONParse("[1,2,3]")).toEqual([1, 2, 3]);
  });

  test("ignores invalid JSON", () => {
    expect(maybeJSONParse("foobar")).toBe(null);
  });
});

describe("logMessageList", () => {
  const logMessageList = CTAN.__get__("logMessageList");

  test("logs error entry", () => {
    const logger = { error: jest.fn() };
    logMessageList(logger, [["ERROR", "Label", "Details"]]);
    expect(logger.error).toHaveBeenCalledWith(
      'CTAN says ["ERROR","Label","Details"]'
    );
  });

  test("logs warning entry", () => {
    const logger = { warning: jest.fn() };
    logMessageList(logger, [["WARNING", "Label", "Details"]]);
    expect(logger.warning).toHaveBeenCalledWith(
      'CTAN says ["WARNING","Label","Details"]'
    );
  });

  test("logs info entry", () => {
    const logger = { info: jest.fn() };
    logMessageList(logger, [["INFO", "Label", "Details"]]);
    expect(logger.info).toHaveBeenCalledWith(
      'CTAN says ["INFO","Label","Details"]'
    );
  });

  test("logs other entry", () => {
    const logger = { info: jest.fn() };
    logMessageList(logger, [["OTHER", "Label", "Details"]]);
    expect(logger.info).toHaveBeenCalledWith(
      'CTAN says ["OTHER","Label","Details"]'
    );
  });

  test("logs many entries", () => {
    const logger = { error: jest.fn() };
    logMessageList(logger, [
      ["ERROR", "Label", "Details"],
      ["ERROR", "Label", "Details"]
    ]);
    expect(logger.error).toHaveBeenCalledTimes(2);
  });
});


test/run.test.js
const YAML = require("yaml");

const run = require("../src/run");

describe("run", () => {
  const ENV_BAK = process.env;

  beforeEach(() => {
    process.exitCode = undefined;
    process.env = { ...ENV_BAK };
  });

  test("failure (missing inputs)", async () => {
    expect(process.exitCode).toBe(undefined);
    await run();
    expect(process.exitCode).toBe(1);
  });

  test("failure (invalid inputs)", async () => {
    process.env["INPUT_VERSION"] = "1.0";
    process.env["INPUT_ACTION"] = "validate";
    process.env["INPUT_FILE_PATH"] = "test/fixtures/htunk.zip";
    process.env["INPUT_FIELDS"] = YAML.stringify({
      pkg: "knuth",
      version: "foobar",
      author: "foobar",
      email: "foobar",
      uploader: "foobar",
      license: "bsd",
      summary: "foobar"
    });
    expect(process.exitCode).toBe(undefined);
    await run();
    expect(process.exitCode).toBe(1);
  });

  test("success", async () => {
    process.env["INPUT_VERSION"] = "1.0";
    process.env["INPUT_ACTION"] = "validate";
    process.env["INPUT_FILE_PATH"] = "test/fixtures/htunk.zip";
    process.env["INPUT_FIELDS"] = YAML.stringify({
      pkg: "htunk",
      version: "foobar",
      author: "foobar",
      email: "foobar",
      uploader: "foobar",
      license: "bsd",
      summary: "foobar"
    });
    expect(process.exitCode).toBe(undefined);
    await run();
    expect(process.exitCode).toBe(undefined);
  });
});


run.js
const core = require("@actions/core");
const fs = require("fs");
const path = require("path");
const YAML = require("yaml");

const CTAN = require("./ctan");

async function run() {
  try {
    core.info("Reading action inputs:");

    const version = core.getInput("version");
    core.info("  Version: " + JSON.stringify(version));

    const action = core.getInput("action");
    core.info("  Action: " + JSON.stringify(action));

    const fields = core.getInput("fields", { required: true });
    const formDataFields = YAML.parse(fields);
    core.info("  Parameters: " + JSON.stringify(formDataFields));

    const filePath = core.getInput("file_path", { required: true });
    const formDataFile = fs.createReadStream(filePath);
    core.info("  Filepath: " + JSON.stringify(path.resolve(formDataFile.path)));

    const formData = { ...formDataFields, file: formDataFile };

    core.info("Starting API request...");

    await CTAN.post({
      version: version,
      action: action,
      formData: formData,
      logger: core
    });
  } catch (error) {
    core.setFailed(error.message);
  }
}

module.exports = run;


src/ctan.js
const request = require("request");
const util = require("util");

const BASE_URL = "https://www.ctan.org/submit";

function post({
  version = undefined,
  action = "validate",
  formData = {},
  logger = console
} = {}) {
  return util
    .promisify(request.post)({
      url: buildUrl({ version: version, action: action }),
      formData: formData
    })
    .then(({ error, statusCode, body }) => {
      postCallback(logger, error, statusCode, body);
    });
}

function buildUrl({ version = undefined, action = "validate" } = {}) {
  if (version === undefined) {
    return [BASE_URL, action].join("/");
  } else {
    return [BASE_URL, version, action].join("/");
  }
}

function postCallback(logger, error, statusCode, body) {
  if (error) throw error;
  logger.info(`CTAN responded with code ${statusCode}`);
  // NOTE: codes expected to return the log as a JSON list are 200, 404 (sometimes), and 409.
  logBody(logger, body);
  // NOTE: documented error codes are only 404, 409, and 500.
  if (statusCode >= 400) throw "CTAN submit failed: see log for details";
}

function logBody(logger, body) {
  const messageList = maybeJSONParse(body);
  if (Array.isArray(messageList)) {
    logMessageList(logger, messageList);
  } else {
    logger.info("CTAN says " + body);
  }
}

function maybeJSONParse(string) {
  try {
    return JSON.parse(string);
  } catch (e) {
    return null;
  }
}

function logMessageList(logger, messageList) {
  messageList.forEach(message => {
    const humanMessage = "CTAN says " + JSON.stringify(message);
    const messageType = message[0];
    switch (messageType) {
      case "ERROR":
        logger.error(humanMessage);
        break;
      case "WARNING":
        logger.warning(humanMessage);
        break;
      default:
        // i.e. "INFO" and everything else
        logger.info(humanMessage);
    }
  });
}

const CTAN = {
  post: post
};

module.exports = CTAN;
